### ISSUE TITLE
Parameter validation failed: Invalid type for parameter ChangeBatch.Changes[1].ResourceRecordSet

### ISSUE THREAD

**Issue by wildinstinct** [aws/aws-cli#2076]
Hi all, I am facing same error as mentioned above. Please help me resolve it. Is the issue related to unicode characters in change-batch json file?

Here is my change resource record set batch file

```json
{
    "Changes": [
        {
            "Action": "CREATE",
            "ResourceRecordSet": {
                "AliasTarget": {
                    "DNSName": "dualstack.internal-react-flfmt-ofrs-elb-dev-blue-617313876.us-east-1.elb.amazonaws.com.",
                    "EvaluateTargetHealth": true,
                    "HostedZoneId": "Z3DZXE0Q79N41H"
                },
                "Failover": "PRIMARY",
                "Name": "reactive-pricing-offers.aws-card-dqa.capitalonegslbex.com.",
                "SetIdentifier": "reactive-pricing-offers-PRIMARY",
                "Type": "A"
            }
        },
        {
            "Action": "UPSERT",
            "ResourceRecordSet": [
                {
                    "AliasTarget": {
                        "DNSName": "dualstack.internal-react-flfmt-ofrs-elb-dev-green-1367806044.us-east-1.elb.amazonaws.com.",
                        "EvaluateTargetHealth": true,
                        "HostedZoneId": "Z3DZXE0Q79N41H"
                    },
                    "Failover": "SECONDARY",
                    "Name": "reactive-pricing-offers.aws-card-dqa.capitalonegslbex.com.",
                    "SetIdentifier": "reactive-pricing-offers-SECONDARY",
                    "Type": "A"
                }
            ]
        }
    ]
}
```

```
Parameter validation failed:
Invalid type for parameter ChangeBatch.Changes[1].ResourceRecordSet, value: [OrderedDict([(u'Type', u'A'), (u'Failover', u'SECONDARY'), (u'AliasTarget', OrderedDict([(u'HostedZoneId', u'Z3DZXE0Q79N41H'), (u'EvaluateTargetHealth', True), (u'DNSName', u'dualstack.internal-react-flfmt-ofrs-elb-dev-green-1367806044.us-east-1.elb.amazonaws.com.')])), (u'SetIdentifier', u'reactive-pricing-offers-SECONDARY'), (u'Name', u'reactive-pricing-offers.aws-card-dqa.capitalonegslbex.com.')])], type: <type 'list'>, valid types: <type 'dict'>
2016-07-24 22:10:26,715 - MainThread - awscli.clidriver - DEBUG - Exiting with rc 255

Parameter validation failed:
Invalid type for parameter ChangeBatch.Changes[1].ResourceRecordSet, value: [OrderedDict([(u'Type', u'A'), (u'Failover', u'SECONDARY'), (u'AliasTarget', OrderedDict([(u'HostedZoneId', u'Z3DZXE0Q79N41H'), (u'EvaluateTargetHealth', True), (u'DNSName', u'dualstack.internal-react-flfmt-ofrs-elb-dev-green-1367806044.us-east-1.elb.amazonaws.com.')])), (u'SetIdentifier', u'reactive-pricing-offers-SECONDARY'), (u'Name', u'reactive-pricing-offers.aws-card-dqa.capitalonegslbex.com.')])], type: <type 'list'>, valid types: <type 'dict'>
```

**Comment by JordonPhillips** [comment]
Your resource record set is a list. Looks like you can just break it out of the list.

**Comment by kyleknap** [comment]
Closing due to inactivity

**Comment by brianpcurran** [comment]
I'm having this issue as well, on both Windows and Linux with awscli installed through pip.

I've created ASCII-encoded files on both Linux and Windows.

`$ file -bi /tmp/route53.json
text/plain; charset=us-ascii`

And using `out-file $route53File -encoding ascii` on Windows.

When calling `aws route53 change-resource-record-sets --change-batch file://foo.json` I get the following output on both Linux and Windows.

`Parameter validation failed:
Invalid type for parameter ChangeBatch.Changes[0].ResourceRecordSet.ResourceRecords, value: OrderedDict([('Value', '172.16.16.1')]), type: <class 'collections.OrderedDict'>, valid types: <class 'list'>, <class 'tuple'>
Invalid type for parameter ChangeBatch.Changes[1].ResourceRecordSet.ResourceRecords, value: OrderedDict([('Value', '172.16.16.2')]), type: <class 'collections.OrderedDict'>, valid types: <class 'list'>, <class 'tuple'>`

This is the traceback I get when using `--debug`.

`Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/awscli/clidriver.py", line 186, in main
    return command_table[parsed_args.command](remaining, parsed_args)
  File "/usr/lib/python2.7/site-packages/awscli/clidriver.py", line 381, in __call__
    return command_table[parsed_args.operation](remaining, parsed_globals)
  File "/usr/lib/python2.7/site-packages/awscli/clidriver.py", line 551, in __call__
    call_parameters, parsed_globals)
  File "/usr/lib/python2.7/site-packages/awscli/clidriver.py", line 675, in invoke
    **parameters)
  File "/usr/lib/python2.7/site-packages/botocore/client.py", line 251, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/usr/lib/python2.7/site-packages/botocore/client.py", line 513, in _make_api_call
    api_params, operation_model, context=request_context)
  File "/usr/lib/python2.7/site-packages/botocore/client.py", line 566, in _convert_to_request_dict
    api_params, operation_model)
  File "/usr/lib/python2.7/site-packages/botocore/validate.py", line 270, in serialize_to_request
    raise ParamValidationError(report=report.generate_report())
ParamValidationError: Parameter validation failed:
Invalid type for parameter ChangeBatch.Changes[0].ResourceRecordSet.ResourceRecords, value: OrderedDict([(u'Value', u'172.20.30.135')]), type: <class 'collections.OrderedDict'>, valid types: <type 'list'>, <type 'tuple'>
Invalid type for parameter ChangeBatch.Changes[1].ResourceRecordSet.ResourceRecords, value: OrderedDict([(u'Value', u'172.20.30.135')]), type: <class 'collections.OrderedDict'>, valid types: <type 'list'>, <type 'tuple'>
2016-10-19 16:09:00,938 - MainThread - awscli.clidriver - DEBUG - Exiting with rc 255`

**Comment by brianpcurran** [comment]
This is the JSON I'm using:

`{
	"Comment": "Adding record for new dev instance.",
	"Changes": [{
		"Action": "UPSERT",
		"ResourceRecordSet": {
			"ResourceRecords": {
				"Value": "172.20.30.159"
			},
			"Type": "A",
			"Name": "*.test.com."
		}
	}, {
		"Action": "UPSERT",
		"ResourceRecordSet": {
			"ResourceRecords": {
				"Value": "172.20.30.159"
			},
			"Type": "A",
			"Name": "test.com."
		}
	}]
}`

**Comment by DivyaNandakumar** [comment]
has any found any solution to this?

**Comment by Melikoth** [comment]
Recently I was woring with Tags and ran into a similar error: Invalid type for parameter TagSpecifications[1].Tags, value: OrderedDict([(u'Key', u'Name'), (u'Value', u'BS1')]), type: <class 'collections.OrderedDict'>, valid types: <type 'list'>, <type 
'tuple'>

In my case I was building my JSON in PowerShell, and my Instance had 3 tags, while the volume only had 1. After some reading it appeared that what was happening was in the case of a single tag, PowerShell was dropping the square brackets and treating it as a single item instead of an array of items containing only one thing.

The multi-tags worked fine when written this way since the 'Tags' was an array.
$TagDataInst = @{
    'ResourceType' = 'instance';
    'Tags' = $TagName,$TagClass,$TagEnvir
    }

But when it failed on the volume tag which was written this way
$TagDataReso = @{
    'ResourceType' = 'volume';
    'Tags' = $TagName
    }

To remedy this, it was required to force the single item into an array like so
$TagDataReso = @{
    'ResourceType' = 'volume';
    'Tags' = @($TagName)
    }

I was then able to run my script without issue.

TL;DR: Something that was supposed to be an array wasn't in square brackets because there was only a single item in the array.  AWS expects the square brackets to be present even if there is only a single item, and it caused the same error you were seeing.

**Comment by apollo17march** [comment]
I was also having this issue, but the error message is pretty self explanatory. "**Invalid type for parameter**" 
both of your  wrong type of data and  valid type is given in the log.

> for example:
> value: OrderedDict([('Value', '172.16.16.1')]), type: <class 'collections.OrderedDict'>, valid types: <class 'list'>, <class 'tuple'>
> in this case, somehow your input is of type OrderedDict while it only accepts  list or tuple
